### PR TITLE
 Example: add second-order viscoacoustic equations

### PR DESCRIPTION
**Pull request by nogueirapeterson**
@lauerami, @speglich, and I developed the second-order forward and adjoint viscoacoustic equations in Devito, for the approaches of SLS (Standard Linear Solid), Ren and Deng. We hope that the Devito team appreciates the contribution.

### CODE REVIEW

File: examples/seismic/viscoacoustic/operators.py
Comment (rhodrin) [inline examples/seismic/viscoacoustic/operators.py:9]: What's the significance behind the `1st` here? Terminology from one of the references?

File: examples/seismic/viscoacoustic/operators.py
Comment (rhodrin) [inline examples/seismic/viscoacoustic/operators.py:65]: Ok, I see, but 1st and 2nd seems a bit vague for the code here. Are you able to be a bit more specific?

File: examples/seismic/viscoacoustic/operators.py
Comment (rhodrin) [inline examples/seismic/viscoacoustic/operators.py:67]: And what exactly do you mean by 'stencil' here? Would something along the lines of 'Implementation of the 2nd order viscoacoustic wave-equation from Bai (2014)' be better?

File: examples/seismic/viscoacoustic/operators.py
Comment (rhodrin) [inline examples/seismic/viscoacoustic/operators.py:78]: `forward = kwargs.get('forward', True)`

File: examples/seismic/viscoacoustic/operators.py
Comment (rhodrin) [inline examples/seismic/viscoacoustic/operators.py:97]: Is there a reason this is not stored directly in `model`?

File: examples/seismic/viscoacoustic/operators.py
Comment (rhodrin) [inline examples/seismic/viscoacoustic/operators.py:180]: Can't this and the above be combined and have a `'case'` `kwarg`?

File: examples/seismic/viscoacoustic/operators.py
Comment (rhodrin) [inline examples/seismic/viscoacoustic/operators.py:99]: Shouldn't this be part of `model`? If it was the appropriate stencil and case could be constricted depending on what fields and `kwargs` are passed right?

File: examples/seismic/viscoacoustic/viscoacoustic_example.py
Comment (rhodrin) [inline examples/seismic/viscoacoustic/viscoacoustic_example.py:42]: What's going on here? This looks stange?

**Comment by rhodrin** [comment]
Thanks again for this contribution. Looks good! I had a quick read through and added a few comments above, will do another pass soonish. Two main comments are:
1. It seems like there's some redundancy and through using a couple of `kwargs` and a few adjustments to `model` some tidying and cutting down can be done.
2. Maybe I missed them, but couldn't see any correctness tests being added (e.g. a test in the example added to ci which checks wave-field and receiver norms)? I only saw some tests checking things actually run?

File: examples/seismic/viscoacoustic/operators.py
Comment (nogueirapeterson) [reply examples/seismic/viscoacoustic/operators.py:65]: We used `stencil` based on available examples. But we agree with you, `order` is better. We will change.

File: examples/seismic/viscoacoustic/operators.py
Comment (nogueirapeterson) [reply examples/seismic/viscoacoustic/operators.py:67]: Yes. We agree. We will change.

File: examples/seismic/viscoacoustic/operators.py
Comment (nogueirapeterson) [reply examples/seismic/viscoacoustic/operators.py:78]: okay. We will fix.

File: examples/seismic/viscoacoustic/operators.py
Comment (nogueirapeterson) [reply examples/seismic/viscoacoustic/operators.py:97]: Yes. We chose this approach based on the available viscoelastic example.

File: examples/seismic/viscoacoustic/operators.py
Comment (nogueirapeterson) [reply examples/seismic/viscoacoustic/operators.py:180]: We understand this approach but based on the available examples we choose to make this case option on equation methods, before calling these methods.

File: examples/seismic/viscoacoustic/operators.py
Comment (nogueirapeterson) [reply examples/seismic/viscoacoustic/operators.py:99]: The variable `b` is already part of the model. The` rho` is computed from `b`.

File: examples/seismic/viscoacoustic/viscoacoustic_example.py
Comment (nogueirapeterson) [reply examples/seismic/viscoacoustic/viscoacoustic_example.py:42]: These values for norm and tolerance are expected because we are considering the second-order SLS equation as a reference. We will fix it to compute all first and second-order equations tests.

**Comment by nogueirapeterson** [comment]
@rhodrin Thanks for the review, we appreciate your comments and learn a lot with them! See you soon.

File: examples/seismic/viscoacoustic/operators.py
Comment (georgebisbas) [inline examples/seismic/viscoacoustic/operators.py:396]: Maybe: Implementation of the 1st and 2nd order viscoacoustic wave-equation from Ren et al. (2014).

File: examples/seismic/viscoacoustic/operators.py
Comment (georgebisbas) [inline examples/seismic/viscoacoustic/operators.py:442]: maybe:
in viscoacoustic media
or
in a viscoacoustic medium

File: examples/seismic/viscoacoustic/operators.py
Comment (georgebisbas) [inline examples/seismic/viscoacoustic/operators.py:453]: kernel: string, optional
Selects...

File: examples/seismic/viscoacoustic/operators.py
Comment (georgebisbas) [inline examples/seismic/viscoacoustic/operators.py:501]: same as before

File: examples/seismic/viscoacoustic/viscoacoustic_example.py
Comment (georgebisbas) [inline examples/seismic/viscoacoustic/viscoacoustic_example.py:34]: leftover?

File: examples/seismic/viscoacoustic/wavesolver.py
Comment (georgebisbas) [inline examples/seismic/viscoacoustic/wavesolver.py:39]: not sure this comment here is helpful?

File: examples/seismic/viscoacoustic/wavesolver.py
Comment (georgebisbas) [inline examples/seismic/viscoacoustic/wavesolver.py:109]: what is the memory variable?

Comment (speglich) [reply examples/seismic/viscoacoustic/wavesolver.py:109]: Hi George, thank you for your review! The R variable is described as memory variable on geophysics domain and is not connected with any computacional expect. This variable contain information about energy loss of stress-strain, and for containing this information is described as a memory variable.

**Comment by rhodrin** [comment]
@nogueirapeterson @speglich Thanks for the latest updates. Would you mind rebasing this on top of the latest master (so that all the tests run) and then myself and @mloubout will take a further look before the end of the week.

**Comment by speglich** [comment]
Resolving these merge issues.

**Comment by speglich** [comment]
Hi @rhodrin! I noticed that the pipeline changed. The new ** Examples with MPI ** test are failing, including in old versions of this PR (We checked this). An initial investigation of the error raised by this test, take us to think about a wrong code generation. We will investigate the error and open an issue, right?

File: examples/seismic/viscoacoustic/operators.py
Comment (mloubout) [inline examples/seismic/viscoacoustic/operators.py:134]: Are these temps really necessary? Is there an issue if you just put
```
 pde_p = 2. * p - damp * p.forward + s * s * vp * vp * div(b * grad((1. + tt) * rho * p))) - \
            s * s * vp * vp * div(b * grad(rho * r.backward))

```

You can still split the expression in a few lines if you want to avoid one long expression but the extra function/vector shulnd't be necessary.

File: examples/seismic/viscoacoustic/operators.py
Comment (mloubout) [inline examples/seismic/viscoacoustic/operators.py:219]: Probably not needed tmps as well.

File: examples/seismic/viscoacoustic/operators.py
Comment (mloubout) [inline examples/seismic/viscoacoustic/operators.py:242]: same

File: examples/seismic/viscoacoustic/operators.py
Comment (mloubout) [inline examples/seismic/viscoacoustic/operators.py:334]: same

File: examples/seismic/viscoacoustic/operators.py
Comment (mloubout) [inline examples/seismic/viscoacoustic/operators.py:353]: same

File: examples/seismic/viscoacoustic/operators.py
Comment (mloubout) [inline examples/seismic/viscoacoustic/operators.py:490]: That factor depends on the PDE, the `dt**2/m` is only valid for 2nd order acoustic WE.

File: examples/seismic/viscoacoustic/operators.py
Comment (mloubout) [inline examples/seismic/viscoacoustic/operators.py:539]: 2nd order acoustic weight `dt**2 / m` pde dependent.

File: examples/seismic/viscoacoustic/wavesolver.py
Comment (mloubout) [inline examples/seismic/viscoacoustic/wavesolver.py:136]: `v=None`

File: examples/seismic/viscoacoustic/wavesolver.py
Comment (mloubout) [inline examples/seismic/viscoacoustic/wavesolver.py:153]: qp, r, b, q missing from docstring.

What's `q` again, pressure? If pressure would not use `q` usually used for source or attenuation and name it something like `pa`. Would also not name particle velocity `u`, that's confusing with most forward using `u` for forward wavefield, maybe `va`?

File: examples/seismic/viscoacoustic/wavesolver.py
Comment (mloubout) [inline examples/seismic/viscoacoustic/wavesolver.py:193]: May want to return particle velocity as well?

File: examples/seismic/viscoacoustic/operators.py
Comment (nogueirapeterson) [reply examples/seismic/viscoacoustic/operators.py:134]: We appreciate your review @mloubout .
These temps were used to avoid the known and described issue about multiple functions on the argument of `grad` operator `Ex: grad((1. + tt) * rho * p)` . 
So we compute each step of the parameters before passing them as arguments.

File: examples/seismic/viscoacoustic/operators.py
Comment (nogueirapeterson) [reply examples/seismic/viscoacoustic/operators.py:490]: Okay. We will fix it.

File: examples/seismic/viscoacoustic/operators.py
Comment (nogueirapeterson) [reply examples/seismic/viscoacoustic/operators.py:539]: Okay. We will fix it.

File: examples/seismic/viscoacoustic/wavesolver.py
Comment (nogueirapeterson) [reply examples/seismic/viscoacoustic/wavesolver.py:136]: okay

File: examples/seismic/viscoacoustic/wavesolver.py
Comment (nogueirapeterson) [reply examples/seismic/viscoacoustic/wavesolver.py:153]: Okay, we will follow your suggestions.

File: examples/seismic/viscoacoustic/operators.py
Comment (mloubout) [reply examples/seismic/viscoacoustic/operators.py:134]: Is it still an issue? I think `grad` is used some places on expressions and should support it. If it fails, could you open an issue with an MFE on it?

Comment (nogueirapeterson) [reply examples/seismic/viscoacoustic/operators.py:134]: Yes, until the day I did the PR, this is still an issue.
You fixed it for the `div`, but not for the `grad` yet.
Yes, I can open an issue. I will do this as soon as possible.

Comment (mloubout) [reply examples/seismic/viscoacoustic/operators.py:134]:
```
In [16]: div(u * grad((1+t) * u * v))                                                                                                                                                
Out[16]: Derivative(u(t, x, y)*Derivative(3*u(t, x, y)*v(x, y), x), x) + Derivative(u(t, x, y)*Derivative(3*u(t, x, y)*v(x, y), y), y)

```

Works fine for expressions this is weird.

Comment (nogueirapeterson) [reply examples/seismic/viscoacoustic/operators.py:134]: I see. But when the stencil is passed to the operator the results produce NAN. Anyway, I will check on Monday with the most updated version of Devito, and if confirmed I will open an issue. Otherwise, I will remove the tmps.

Comment (mloubout) [reply examples/seismic/viscoacoustic/operators.py:134]: >passed to the operator the results produce NAN.

Now that's something I can get behind, may need some parametrizable grad/div  to keep it nice and stable (i.e like how you can do `u.dx(x0=x + dx)` but for grad/div. I'll see if I can find time to play around with your PR and file a patch for it but may take a little bit so can leave the implementation like that.

Comment (nogueirapeterson) [reply examples/seismic/viscoacoustic/operators.py:134]: Got it. I will keep implementing it that way.
Thanks for the clarification.
On Monday I will make the suggested corrections.

File: examples/seismic/viscoacoustic/operators.py
Comment (mloubout) [inline examples/seismic/viscoacoustic/operators.py:158]: Since these are local tmps, you may only need `VectorFunction, Function` and not time dependent, same other places

**Comment by speglich** [comment]
Hi! We updated the PR removing unnecessary/duplicated functions and increase the tolerance to viscoacoustic tests.

File: examples/seismic/viscoacoustic/operators.py
Comment (speglich) [reply examples/seismic/viscoacoustic/operators.py:160]: Hi @mloubout, we understood your review and we tried this approach. When we replace `TimeFunction` by `Function` we are experiencing seg fault with the operator 😭.

Comment (mloubout) [reply examples/seismic/viscoacoustic/operators.py:160]: That's quite odd :O

File: examples/seismic/viscoacoustic/operators.py
Comment (nogueirapeterson) [inline examples/seismic/viscoacoustic/operators.py:366]: @FabioLuporini  these auxiliary functions are temporaries.
Some of them are like `TimeFunction` / `VectorTimeFunction`, however it would not be necessary. Just `Function` / `VectorFunction` should be enough. But the last one generates a segmentation fault.

**Comment by FabioLuporini** [comment]
CI is currently failing because of issue #1459 . I've assigned it to me

**Comment by Leitevmd** [comment]
I've opened #1483 which I think is the only problem right now (after the last commit) for this PR. I've have already proposed a solution. Just waiting for approval to submit a new PR.

**Comment by FabioLuporini** [comment]
I've just merged master into this branch since the random OSX failure should have now disappeared. Let's see

**Comment by nogueirapeterson** [comment]
thanks, @FabioLuporini.  Now disappeared the OSX failure. All tests are green!

**Review by mloubout — COMMENTED** [review]
Really nice, glad all working without temporaries now. Did you check the result look ok? May need to use the shift in div/grad i.e `div(grad(expr, shift=.5), shift=-.5)` to get nice results

File: examples/seismic/viscoacoustic/operators.py
Comment (FabioLuporini) [reply examples/seismic/viscoacoustic/operators.py:25]: > glad all working without temporaries now. 

can we assert the expected wavefield norms in the notebooks?

File: tests/test_adjoint.py
Comment (FabioLuporini) [inline tests/test_adjoint.py:47]: @mloubout @rhodrin now that I note this... that's a lot of tests added! Do you think we actually need all of them?

File: examples/seismic/viscoacoustic/operators.py
Comment (FabioLuporini) [reply examples/seismic/viscoacoustic/operators.py:25]: just noted that we're already testing via `test_viscoacoustic`, so we should be good

**Comment by nogueirapeterson** [comment]
> Really nice, glad all working without temporaries now. Did you check the result look ok? May need to use the shift in div/grad i.e `div(grad(expr, shift=.5), shift=-.5)` to get nice results

After removing the tmps, I didn't check the results. But all the tests worked well (stability, amplitude, and adjoint).
Anyway, I'm going to check for security.

**Comment by nogueirapeterson** [comment]
> > Really nice, glad all working without temporaries now. Did you check the result look ok? May need to use the shift in div/grad i.e `div(grad(expr, shift=.5), shift=-.5)` to get nice results
> 
> After removing the tmps, I didn't check the results. But all the tests worked well (stability, amplitude, and adjoint).
> Anyway, I'm going to check for security.

@mloubout 
@FabioLuporini 
We did three experiments using the 2nd acoustic wave equation. First without any modification, with temporaries, after without temporaries, and finally, using the shift parameter:
shots

Attached with this comment, we leave a script explaining how we did these experiments.
[test_wo_tmps.txt](https://github.com/devitocodes/devito/files/5584273/test_wo_tmps.txt)

**Comment by FabioLuporini** [comment]
@nogueirapeterson if your tests were still passing despite the images being "wrong" or "flawed" then it _also_ means that the provided tests are not sufficient. Does it make sense @mloubout or am I missing something?

**Comment by nogueirapeterson** [comment]
> @nogueirapeterson if your tests were still passing despite the images being "wrong" or "flawed" then it _also_ means that the provided tests are not sufficient. Does it make sense @mloubout or am I missing something?

The existing tests analyze stability, amplitude, and the adjoint operation. Removing the temps did not directly affect these tests. 
However, it is producing a noise that I guess is related to the derivatives on a staggered grid.

**Comment by nogueirapeterson** [comment]
@FabioLuporini and @mloubout We are waiting for PR #1551, once it is approved, possibly this PR will be completed.

**Review by FabioLuporini — APPROVED** [review]
Great, all good now

**Review by rhodrin — APPROVED** [review]

**Pull request merged**
